cns/restserver/ipampoolmonitor.go

```diff
@@ -0,0 +1,138 @@
+)
+
+var (
+	increasePoolSize = 1
+	decreasePoolSize = -1
+	doNothing        = 0
+)
+
+type IPAMPoolMonitor interface {
+	Start() error
+	UpdatePoolLimitsTransacted(batchSize int, requestThreshold float64, releaseThreshold float64)
+}
+
+type CNSIPAMPoolMonitor struct {
+	initialized bool
+
+	cns *HTTPRestService
+	rc  requestcontroller.RequestController
+
+	batchSize        int
+	requestThreshold float64
+	releaseThreshold float64
+	minimumFreeIps   int
+	maximumFreeIps   int
+	sync.RWMutex
+}
+
+func NewCNSIPAMPoolMonitor(cnsService *HTTPRestService, requestController requestcontroller.RequestController) *CNSIPAMPoolMonitor {
+	return &CNSIPAMPoolMonitor{
+		initialized: false,
+		cns:         cnsService,
+		rc:          requestController,
+	}
+}
+
+// TODO: add looping and cancellation to this, and add to CNS MAIN
+func (pm *CNSIPAMPoolMonitor) Start() error {
+
+	if pm.initialized {
+		availableIPConfigs := pm.cns.GetAvailableIPConfigs()
+		rebatchAction := pm.checkForResize(len(availableIPConfigs))
+		switch rebatchAction {
+		case increasePoolSize:
+			return pm.increasePoolSize()
+		case decreasePoolSize:
+			return pm.decreasePoolSize()
+		}
+	}
+	return nil
+}
+
+// UpdatePoolLimitsTransacted called by request controller on reconcile to set the batch size limits
+func (pm *CNSIPAMPoolMonitor) UpdatePoolLimitsTransacted(batchSize int, requestThreshold float64, releaseThreshold float64) {
+
+	pm.Lock()
+	defer pm.Unlock()
+	pm.batchSize = batchSize
+	pm.requestThreshold = requestThreshold
+	pm.releaseThreshold = releaseThreshold
+
+	//TODO: rounding
+	pm.minimumFreeIps = int(float64(pm.batchSize) * requestThreshold)
+	pm.maximumFreeIps = int(float64(pm.batchSize) * releaseThreshold)
+	pm.initialized = true
+}
+
+func (pm *CNSIPAMPoolMonitor) checkForResize(freeIPConfigCount int) int {
+	switch {
+	// pod count is increasing
+	case freeIPConfigCount < pm.minimumFreeIps:
+		logger.Printf("Number of free IP's (%d) < minimum free IPs (%d), request batch increase\n", freeIPConfigCount, pm.minimumFreeIps)
+		return increasePoolSize
+
+	// pod count is decreasing
+	case freeIPConfigCount > pm.maximumFreeIps:
+		logger.Printf("Number of free IP's (%d) > maximum free IPs (%d), request batch decrease\n", freeIPConfigCount, pm.maximumFreeIps)
+		return decreasePoolSize
+	}
+	return doNothing
+}
+
+func (pm *CNSIPAMPoolMonitor) increasePoolSize() error {
+	increaseIPCount := len(pm.cns.PodIPConfigState) + pm.batchSize
+
+	// pass nil map to CNStoCRDSpec because we don't want to modify the to be deleted ipconfigs
+	spec, err := CNSToCRDSpec(nil, increaseIPCount)
+	if err != nil {
+		return err
+	}
+
+	return pm.rc.UpdateCRDSpec(context.Background(), spec)
+}
+
+func (pm *CNSIPAMPoolMonitor) decreasePoolSize() error {
+
+	// TODO: Better handling here, negatives?
+	decreaseIPCount := len(pm.cns.PodIPConfigState) - pm.batchSize
+
+	// mark n number of IP's as pending
+	pendingIPAddresses, err := pm.cns.MarkIPsAsPendingTransacted(decreaseIPCount)
```

Comment:
Can you expose a method in internalapi.go than calling a TransactedAPI directly? If this poolMonitor is outside the RestServer package, then we cant call Transacted APIs directl. Also it will help to expose the same as debug API from client.

Reply:
it should be  pm.cns.MarkIPsAsPendingTransacted(pm.batchsize)